lasagne/layers/base.py

```diff
@@ -23,8 +23,10 @@ class Layer(object):
     network's output :class:`Layer` instance doubles as a handle to the full
     network.
     """
-    def __init__(self, input_layer):
-        self.input_layer = input_layer
+    def __init__(self, input_layer, name=None):
+        if input_layer is not None:
+            self.input_layer = input_layer
```

Comment:
Duplicating the logic for `name` in `InputLayer` kind of smells imo, which is why I modified it so it would call the constructor now. But clearly that isn't an ideal situation either.

Either way catching `input_layer is None` and raising an exception is not something we can do in the `Layer` base class, because the `InputLayer` either doesn't have this property, or it should be `None` (depending on which solution we go for).

But we can't really go around and put that check in every `Layer` subclass either. An intermediate subclass for layers with an input_layer is a possibility, but that complicates things quite a bit and I'm not sure if we want that.